internal/configv2/configv2.go

```diff
@@ -0,0 +1,409 @@
+		if err := sub.Unmarshal(receiverCfg); err != nil {
+			return nil, &configError{
+				code: errUnmarshalError,
+				msg:  fmt.Sprintf("error reading settings for receiver type %q: %v", typeStr, err),
+			}
+		}
+		receivers = append(receivers, receiverCfg)
+	}
+
+	return receivers, nil
+}
+
+func loadExporters(v *viper.Viper) ([]models.ExporterCfg, error) {
+	// Get the list of all "exporters" items from config source
+	subs, err := getViperSubSequenceOfMaps(v, models.ExportersKeyName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over exporters and create a config for each
+	exporters := make([]models.ExporterCfg, 0)
+	for _, sub := range subs {
+		// Find exporter factory based on "type" that we read from config source
+		typeStr := sub.GetString(models.TypeKeyName)
+		if typeStr == "" {
+			return nil, &configError{code: errMissingExporterType, msg: "missing exporter type"}
+		}
+
+		factory := models.GetExporterFactory(typeStr)
+		if factory == nil {
+			return nil, &configError{code: errUnknownExporterType, msg: fmt.Sprintf("unknown exporter type %q", typeStr)}
+		}
+
+		// Create the default config for this exporter
+		exporterCfg := factory.CreateDefaultConfig()
+
+		// Now that the default config struct is created we can Unmarshal into it
+		// and it will apply user-defined config on top of the default.
+		if err := sub.Unmarshal(exporterCfg); err != nil {
+			return nil, &configError{
+				code: errUnmarshalError,
+				msg:  fmt.Sprintf("error reading settings for exporter type %q: %v", typeStr, err),
+			}
+		}
+		exporters = append(exporters, exporterCfg)
+	}
+
+	return exporters, nil
+}
+
+func loadOptions(v *viper.Viper) ([]models.OptionCfg, error) {
+	// Get the list of all "options" items from config source
+	subs, err := getViperSubSequenceOfMaps(v, models.OptionsKeyName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over options and create a config for each
+	options := make([]models.OptionCfg, 0)
+	for _, sub := range subs {
+		// Find option factory based on "type" that we read from config source
+		typeStr := sub.GetString(models.TypeKeyName)
+		if typeStr == "" {
+			return nil, &configError{code: errMissingOptionType, msg: "missing option type"}
+		}
+
+		factory := models.GetOptionFactory(typeStr)
+		if factory == nil {
+			return nil, &configError{code: errUnknownOptionType, msg: fmt.Sprintf("unknown option type %q", typeStr)}
+		}
+
+		// Create the default config for this options
+		optionCfg := factory.CreateDefaultConfig()
+
+		// Now that the default config struct is created we can Unmarshal into it
+		// and it will apply user-defined config on top of the default.
+		if err := sub.Unmarshal(optionCfg); err != nil {
+			return nil, &configError{
+				code: errUnmarshalError,
+				msg:  fmt.Sprintf("error reading settings for option type %q: %v", typeStr, err),
+			}
+		}
+		options = append(options, optionCfg)
+	}
+
+	return options, nil
+}
+
+func validateConfig(cfg *models.ConfigV2) error {
```

Comment:
Nit: there could be other more subtle bugs in the config. For example if we have both a Zipkin receiver and a Zipkin exporter, their endpoint should be different, otherwise we'll enter an infinite self-loop. Consider adding a TODO here.

Reply:
Good point. Added a comment.